### PR TITLE
Make many helpers, factories and tables static in Microsoft.CSharp

### PR DESCRIPTION
**Pull request by JonHanna**
Microsoft.CSharp has several objects that are heavily used throughout the library that have circular dependencies on each other, as well as several other classes depending on them. Use of them hence requires either a field in each object that will use them, or to follow chains of such fields. Prior to #25191 it was necessary for them to be instances so that they could be replaced with new versions. This no longer being necessary it's possible to make them all static.

At the same time, the related `RuntimeBinder`, `ExpressionBinder` and `BindingContext` classes are relatively heavy, mostly due to these dependencies. They are hence created once unless a cast must be attempted with a different `BindingContext`, and `BindingContext`'s state is overwritten for each operation. Removing these dependencies makes these very light, allowing them to be `readonly struct`s that are created for each operation.

* Have `TypeArray` in charge of the empty `TypeArray` singleton

* Make `TypeArray` allocation a static responsibility of `TypeArray`

* Remove some orphaned members.

Mostly `BSYMMGR`-returning members that had been used to allocate `TypeArrays`

* Make members static where possible.

Those that can already be easily made static. Includes removing some members that this makes no longer necessary.

* Remove unused parameters

Since some are of the types we'll be making static, and such parameters will have to go anyway.

* Make `TypeTable` static

And more conventional names within `TypeTable`.

* More `TypeManager` state & methods to static

* Make `SYMTBL.Key` an immutable struct.

* `SYMTBL` completely static and renamed to `SymbolStore`

More conventional name, and doesn't clash with other `SymbolTable` class.

* `GetWinRTCollectionIfacesAll` to property.

* `SymFactory` static class.

* Remove `PredefinedTypes._symbolManager` field

Only used in assertion.

* Remove assert comparison with this in `GetAggregate`

Assert is the only reason to not be static, allows a cascade of more changes to static.

* Make `InternalsVisibleTo` check static.

* Make cycle of instance methods static

Take quite a large set of methods which must be instance because of cyclic instance dependencies, and make them all static.

Removes dependency in `AggregateSymbol` and `NullableType` on `TypeManager` instances.

* Replace `GetAggTypeSym` with virtual method on `CType`.

Merged with existing `GetAts` on `NullableType`. Comment says this would be nice, and indeed it would, and its easy now.

* Make those members that can now be static, static.

* Remove orphaned members and fields.

* `PredefinedMembers` entirely static.

* Remove `UserStringBuilder.m_buildingInProgress`

Only used in asserts, just assert on whether `m_strBuilder` is null or not.

Also rename `m_strBuilder` to `_strBuilder` to be more conventional.

* Favour returning over out in `UserStringBuilder`

And favour clearing `StringBuilder` over creating another.

* Make `ErrorHandling` completely static.

Creating `UserStringBuilder` instances on demand.

* `TypeManager` static.

* Make `CNullable` methods static and merge directly into `ExpressionBinder`

Don't cache method and property, as they're already cached.

* Don't preload types.

Either they'll be loaded soon anyway, or else it's a wasted effort.

* Move `CompareTypes` into Better.cs

* Remove `BSYMMGR.LookupAggMember`

Just call into `SymbolStore.LookupSym`

* Move `LookupNextSym` logic into `Symbol`

* Remove `BSYMMGR` and `GlobalSymbolContext`

No longer any use.

* `ExprFactory` static

* `SymbolLoader` static

* `CSemanticChecker` static

* Don't pre-emptively load `System.Object`

Will be loaded as a base of the first type encountered anyway.

* `SymbolTable` static

* More methods can be made static

* Remove orphaned method.

* Move methods for getting `MemberInfo`s from expressions into those expressions

* Replace `ExpressionTreeCallRewriter.GetObject` with virtual property in `Expr`

* `TypeArray` tidy-up.

* Pass `ExpressionBinder` to operator delegates, and make them static.

Allows array to be created once statically, rather than on each instance construction.

* Create `RuntimeBinder`, `ExpressionBinder` and `BindingContext` per operation

These are now very lightweight, so it's simpler to create them anew each time rather than overwriting the values in the `BindingContext`.

* Make `RuntimeBinder`, `ExpressionBinder` and `BindingContext` structs

No need to allocate on the heap.

* Remove `AggregateDeclaration`

We only ever care about the `AggregateSymbol` it relates to.

(The two aren't really separate in dynamic code, where there aren't really declarations).

* Remove `IsChecked` and `CallingContext` from C# binders.

Now not used as they are passed to `RuntimeBinder` on construction, except for `CallingContext` remaining in `CSharpInvokeMemberBinder` for use in deferring.

* Make `RecordBinOpSigFromArgs` static.

A last member that can be made static.

CC @VSadov @OmarTawfik

### CODE REVIEW

**Comment by danmosemsft** [comment]
@JonHanna thanks, just curious why you make so many contributions to this library specifically?

**Comment by JonHanna** [comment]
@danmosemsft because I'm hooked, and the more I do the more I see to do (my focus started with one or two things with Linq that had annoyed me when it was only closed source, then seeing more to do there led me to things in Linq.Expressions, which is fascinating in defining one language and then producing two others [since the interpreter's output is a language in itself], which led me to this).
But also I think I'm drawn to it more than I benefit from as a user because there's both so much I like about this library (the very fact that it's doing live dynamic binding on arbitrary objects is pretty awesome) and so much I dislike (the ported origin leaving redundancies, and a lot of C# code that "smells" of C++, and which can be confusing when trying to do a necessary fix) and that combination draws it to me to it where code I only admired I'd find nothing to contribute to, and code I only disliked I wouldn't care about to contribute to in the first place.

File: src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Tree/Constant.cs [hunk 1]
```diff
@@ -44,5 +45,68 @@ public long Int64Value
...
+        {
+            get
+            {
+                if (Type is NullType)
+                {
+                    return null;
+                }
+
+                object objval;
+                switch (System.Type.GetTypeCode(Type.AssociatedSystemType))
```
Comment (JonHanna) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Tree/Constant.cs]: @OmarTawfik moving this switch to `ExprConstant` is what I suggested would be ideal in reply to your comment at https://github.com/dotnet/corefx/pull/26414#discussion_r163115728 but for requiring dependencies on `TypeManager` that in other types' equivalent that would go away if `TypeManager` was static.

**Comment by danmosemsft** [comment]
@jonhanna ha, those are good reasons!

**Comment by JonHanna** [comment]
Test Windows x86 Release Build

https://mc.dot.net/#/user/JonHanna/pr~2Fjenkins~2Fdotnet~2Fcorefx~2Fmaster~2F/test~2Ffunctional~2Fcli~2F/728bfea0b256b8bc95e259cf7067b0b109df6ea1

File: src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/BindingContext.cs [hunk 2]
```diff
@@ -2,45 +2,30 @@
...
-
 namespace Microsoft.CSharp.RuntimeBinder.Semantics
 {
     // ----------------------------------------------------------------------------
     // This is the interface for the BindingContext, which is
     // consumed by the StatementBinder.
     // ----------------------------------------------------------------------------
 
-    internal sealed class BindingContext
+    internal readonly struct BindingContext
```
Comment (VSadov) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/BindingContext.cs]: Very nice change!!!

File: src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ICSharpBinder.cs [hunk 3]
```diff
@@ -10,19 +10,19 @@ namespace Microsoft.CSharp.RuntimeBinder
...
         // This is true for any binder that is eligible to take value type receiver
         // objects as a ref (for mutable operations). Such as calls ("v.M(d)"),
         // and indexers ("v[d] = v[d]"). Note that properties are not here because they
         // are only dispatched dynamically when the receiver is dynamic, and hence boxed.
         bool IsBinderThatCanHaveRefReceiver { get; }
 
-        void PopulateSymbolTableWithName(SymbolTable symbolTable, Type callingType, ArgumentObject[] arguments);
+        void PopulateSymbolTableWithName(Type callingType, ArgumentObject[] arguments);
 
         Expr DispatchPayload(RuntimeBinder runtimeBinder, ArgumentObject[] arguments, LocalVariableSymbol[] locals);
```
Comment (VSadov) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ICSharpBinder.cs]: Just wonder - how large is the `RuntimeBinder` now? Should it be passed as an `in` parameter?   (not in this change, just a thought..)

Comment (JonHanna) [reply src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ICSharpBinder.cs]: It's a reference and a bool

Comment (VSadov) [reply src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ICSharpBinder.cs]: Yes. It contains other structs but eventually it is indeed just the BindingContext. 
Such small struct is unlikely to benefit from `in`. Actually with byval there are more chances that JIT can do something smart - promote into scalars, pass in registers, etc..

---
In reply to: [168586125](https://github.com/dotnet/corefx/pull/27044#discussion_r168586125) [](ancestors = 168586125)

Comment (JonHanna) [reply src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ICSharpBinder.cs]: That was my reasoning, and it's not passed often enough per operation to make it worth profiling just to see if something unexpected makes `in` pay off, IMO.

**Pull request merged**